packages/ledgerstate/output.go

```diff
@@ -0,0 +1,1031 @@
+	outputMetadata = &OutputMetadata{}
+	if outputMetadata.id, err = OutputIDFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse OutputID: %w", err)
+		return
+	}
+	if outputMetadata.branchID, err = BranchIDFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse BranchID: %w", err)
+		return
+	}
+	if outputMetadata.solid, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse solid flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	if outputMetadata.solidificationTime, err = marshalUtil.ReadTime(); err != nil {
+		err = xerrors.Errorf("failed to parse solidification time (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	consumerCount, err := marshalUtil.ReadUint64()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse consumer count (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	outputMetadata.consumerCount = int(consumerCount)
+	if outputMetadata.firstConsumer, err = TransactionIDFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse first consumer: %w", err)
+		return
+	}
+	if outputMetadata.preferred, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse preferred flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	if outputMetadata.liked, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse liked flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	if outputMetadata.finalized, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse finalized flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	if outputMetadata.confirmed, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse confirmed flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	if outputMetadata.rejected, err = marshalUtil.ReadBool(); err != nil {
+		err = xerrors.Errorf("failed to parse rejected flag (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+
+	return
+}
+
+// OutputMetadataFromObjectStorage restores an OutputMetadata object that was stored in the ObjectStorage.
+func OutputMetadataFromObjectStorage(key []byte, data []byte) (outputMetadata *OutputMetadata, err error) {
+	if outputMetadata, _, err = OutputMetadataFromBytes(byteutils.ConcatBytes(key, data)); err != nil {
+		err = xerrors.Errorf("failed to parse OutputMetadata from bytes: %w", err)
+		return
+	}
+
+	return
+}
+
+// ID returns the OutputID of the Output that the OutputMetadata belongs to.
+func (o *OutputMetadata) ID() OutputID {
+	return o.id
+}
+
+// BranchID returns the identifier of the Branch that the Output was booked in.
+func (o *OutputMetadata) BranchID() BranchID {
+	o.branchIDMutex.RLock()
+	defer o.branchIDMutex.RUnlock()
+
+	return o.branchID
+}
+
+// SetBranchID sets the identifier of the Branch that the Output was booked in.
+func (o *OutputMetadata) SetBranchID(branchID BranchID) (modified bool) {
```

Comment:
I'd simplify all `SetX` methods to only acquire the write lock. It becomes less verbose, easier to read and you safe the replicated `if`.

Reply:
Done